docs/design/schemas.md

```diff
@@ -0,0 +1,180 @@
+# Provider, resource, and data source schemas
+
+Providers, resources, and data sources have schemas. This document explores the options for how they relate to each other, and the types provider developers use to define them.
+
+This problem is fairly symmetrical with respect to providers, resources, and data sources, so we choose to focus on resource schemas without loss of generality.
+
+## Relationship of schema to resource
+
+In `helper/schema`, `Resource` is a struct type with a field `Schema`, whose type is `map[string]*Schema`.
+
+In the new framework, `framework.Resource` may be a struct type or interface: see the [Structs and Interfaces](./structs-interfaces.md) design doc. 
+
+Since the schema is defined by the provider code but must be accessible from the framework code, this means that if `framework.Resource` is a struct, the schema can either be a field (`framework.Resource.Schema`), or a method (`framework.Resource.GetSchema()`), i.e. we could have
+```go
+type Resource struct {
+  Schema *tfprotov6.Schema
+}
+```
+
+or
+```go
+type Resource struct {
+  GetSchema func(context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic)
+}
+```
+
+If `framework.Resource` is an interface, the only option is a method (`framework.Resource.GetSchema()`), since the actual resource struct will be a provider-defined type:
```

Comment:
I think this isn't _strictly_ true. For example, we could do something like this:

```go
type Resource interface {
  // whatever whatever whatever
  GetSchema(ctx context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic)
}

type Schema map[string]*tfprotov6.Schema

func (s Schema) GetSchema(_ context.Context) (*tfprotov6.Schema, []*tfprotov6.Diagnostic) {
  // implement a conversion function here that converts from s to *tfprotov6.Schema
}
```

Then users could do something like this:

```go
type myResource struct {
  framework.Schema
}

func NewMyResource() framework.Resource {
  return myResource{
    "foo": *tfprotov6.Schema{},
    "bar": *tfprotov6.Schema{},
  }
}
```

Which lets them use a method _or_ a struct implementation. Similarly, we could make `GetSchema` an unexported method in the interface (`getSchema`) which would then _require_ them to use our struct implementation.

Whether any of that is a good idea or not I think is open for discussion, but I do think it's an option that exists for us.

Reply:
Yes, great point. Added this option in.